watertap/core/zero_order_costing.py

```diff
@@ -3032,6 +3010,229 @@ def cost_vfa_recovery(blk):
             blk.unit_model.heat_consumption[t0], "heat"
         )
 
+    def cost_magprex(blk):
+        """
+        Method for costing Magprex reactor unit.
+        """
+        t0 = blk.flowsheet().time.first()
+
+        # Get parameter dict from database
+        parameter_dict = blk.unit_model.config.database.get_unit_operation_parameters(
+            blk.unit_model._tech_type, subtype=blk.unit_model.config.process_subtype
+        )
+
+        # Get costing parameter sub-block for this technology
+        HRT, size_cost = _get_tech_parameters(
+            blk,
+            parameter_dict,
+            blk.unit_model.config.process_subtype,
+            ["HRT", "sizing_cost"],
+        )
+
+        # Add cost variable and constraint
+        blk.capital_cost = pyo.Var(
+            initialize=1,
+            units=blk.config.flowsheet_costing_block.base_currency,
+            bounds=(0, None),
+            doc="Capital cost of unit operation",
+        )
+
+        expr = pyo.units.convert(
+            blk.unit_model.properties_in[t0].flow_vol * HRT * size_cost,
+            to_units=blk.config.flowsheet_costing_block.base_currency,
+        )
+
+        # Determine if a costing factor is required
+        factor = parameter_dict["capital_cost"]["cost_factor"]
+
+        if factor == "TPEC":
+            expr *= blk.config.flowsheet_costing_block.TPEC
+        elif factor == "TIC":
+            expr *= blk.config.flowsheet_costing_block.TIC
+
+        blk.capital_cost_constraint = pyo.Constraint(expr=blk.capital_cost == expr)
+
+        # Register flows
+        blk.config.flowsheet_costing_block.cost_flow(
+            blk.unit_model.electricity[t0], "electricity"
+        )
+        blk.config.flowsheet_costing_block.cost_flow(
+            blk.unit_model.MgCl2_flowrate[t0], "magnesium_chloride"
```

Comment:
You've added magnesium chloride as a registered flow in the costing for Magprex, but magnesium chloride isn't actually a component in the flowsheet. Instead, you have `"solute_list": ["phosphates", "struvite"]`. So then does adding magnesium chloride as a registered flow actually have any effect?

Reply:
From my understanding, MgCl2 is just a coagulant, so I didn't think it had to included as a component in the solute list. It doesn't have any effect on the flowsheet outside of the costing information (the cost of the MgCl2).